crates/ty_python_semantic/src/types/builder.rs

```diff
@@ -451,10 +476,10 @@ impl<'db> UnionBuilder<'db> {
                     }
 
                     if ty.is_equivalent_to(self.db, element_type)
```

Comment:
I'm surprised the equivalence check here is safe if the subtype check isn't. Doesn't the `is_equivalent_to` implementation also involve evaluating the inner value?

It seems like it should be fine to also avoid simplifying equivalent types out of unions -- we could probably replace this with a naive `==` check and be fine?

Reply:
In the latest version of the PR, I no longer need the new UnionStrategy, so I've removed it; we can add it back in and consider it separately for UX or perf (not cycle) reasons.

It's no longer needed because now `TupleType::into_class_type` is a Salsa query, and cycle handling on that query means we no longer need to avoid the union simplification.